src/Microsoft.Azure.Batch.SoftwareEntitlement/NodeEntitlements.cs

```diff
@@ -118,47 +121,73 @@ public NodeEntitlements UntilInstant(DateTimeOffset notAfter)
+        public NodeEntitlements WithApplications(IEnumerable<string> applications)
         {
-            if (application == null)
+            if (applications == null)
             {
-                throw new ArgumentNullException(nameof(application));
+                throw new ArgumentNullException(nameof(applications));
             }
 
-            return new NodeEntitlements(this, applications: Applications.Add(application.Trim()));
+            applications = applications.Select(appId => appId.Trim());
+            return new NodeEntitlements(this, applications: ImmutableHashSet.CreateRange(applications));
         }
 
         /// <summary>
-        /// Specify the IPAddress of the entitled machine
+        /// Specify the IPAddresses of the entitled machine
         /// </summary>
-        /// <param name="address">IP Address of the machine to run the entitled application(s).</param>
+        /// <param name="address">IP Addresses of the machine to run the entitled application(s).</param>
         /// <returns>A new entitlement</returns>
-        public NodeEntitlements AddIpAddress(IPAddress address)
+        public NodeEntitlements WithIpAddresses(params IPAddress[] ipAddresses)
         {
-            if (address == null)
+            if (ipAddresses == null)
             {
-                throw new ArgumentNullException(nameof(address));
+                throw new ArgumentNullException(nameof(ipAddresses));
             }
 
-            return new NodeEntitlements(this, addresses: IpAddresses.Add(address));
+            return WithIpAddresses((IEnumerable<IPAddress>)ipAddresses);
         }
 
         /// <summary>
-        /// Specify the entitlement Id to use
+        /// Specify the list of IPAddresses of the entitled machine
         /// </summary>
-        /// <param name="identifier">The entitlement identifier to use for correlating activity.</param>
+        /// <param name="address">IP Addresses of the machine to run the entitled application(s).</param>
         /// <returns>A new entitlement</returns>
-        public NodeEntitlements WithIdentifier(string identifier)
+        public NodeEntitlements WithIpAddresses(IEnumerable<IPAddress> ipAddresses)
         {
-            if (string.IsNullOrEmpty(identifier))
+            if (ipAddresses == null)
             {
-                throw new ArgumentException("Expect to have an identifier", nameof(identifier));
+                throw new ArgumentNullException(nameof(ipAddresses));
             }
 
+            return new NodeEntitlements(this, addresses: ImmutableHashSet.CreateRange(ipAddresses));
+        }
+
+        /// <summary>
+        /// Specify the entitlement Id to use
+        /// </summary>
+        /// <param name="identifier">The entitlement identifier to use for correlating activity.</param>
+        /// <returns>A new entitlement</returns>
+        public NodeEntitlements WithIdentifier(string identifier)
+        {
```

Comment:
Intentionally removed the guard clause here. Even though we know we should "always" (except for unit tests) have this value, it's really the responsibility of the `IEntitlementPropertyProvider` to do that validation (correctly implementing guard clauses here requires knowledge of the provider implementations). Throwing here is not a reliable means of preventing a null/empty value anyway, since you can achieve that by omitting the `WithIdentifier` call. And it makes unit testing a lot easier if we can specify nulls here.